### PR TITLE
Add _expand_user to _stringify_path as we need to allow tilde - ~, to expand to full path

### LINKED ISSUE #23473: Use expanduser when writing dataframes to files?
#### Code Sample, a copy-pastable example if possible

The following `to_XXX` do not work:
```python
df = pd.DataFrame([[1,1],[2,2]])
df.to_pickle('~/df.pkl')
df.to_json('~/df.json')
 # etc.

```
#### Problem description

`to_XXX` do not work when the path argument includes a `~`.  This may be fine, but it is strange that dataframe *reading* from files does support `~`.    i.e. `pd.read_json('~/df.json')` works fine.

#### Expected Output
I expect `to_json` and friends to support expanding the home directory.

Perhaps the fix is to add `expanduser` to https://github.com/pandas-dev/pandas/blob/0d6cb3c21d1305d75878eee7a470bf463b6754bd/pandas/io/common.py#L121
as we have in https://github.com/pandas-dev/pandas/blob/0d6cb3c21d1305d75878eee7a470bf463b6754bd/pandas/io/common.py#L229

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.6.final.0
python-bits: 64
OS: Darwin
OS-release: 17.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: 3.9.1
pip: 9.0.1
setuptools: 40.4.1
Cython: 0.27.3
numpy: 1.15.2
scipy: 1.1.0
pyarrow: 0.10.0
xarray: None
IPython: 5.3.0
sphinx: 1.5.1
patsy: 0.4.1
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.5
feather: None
matplotlib: 2.2.2
openpyxl: 2.4.1
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 4.1.1
bs4: 4.5.3
html5lib: 0.9999999
sqlalchemy: 1.1.15
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: 0.1.5
fastparquet: 0.1.3
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by srinivasreddy**
- [x] closes #23473 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
we have some tests in io/test_common.py ; do these need to be expanded?

**Comment by jreback** [comment]
ok, can you add a whatsnew note

**Review by WillAyd — CHANGES_REQUESTED** [review]
I actually would prefer adding a test case here which could be parametrized like say `test_read_non_existant` within the same module.

File: pandas/tests/io/test_common.py [hunk 1]
```diff
@@ -150,6 +150,26 @@ def test_read_non_existant(self, reader, module, error_class, fn_ext):
...
+        (pd.read_json, 'os', ValueError, 'json'),
+        (pd.read_msgpack, 'os', ValueError, 'mp'),
+        (pd.read_pickle, 'os', FileNotFoundError, 'pickle'),
+    ])
+    def test_read_non_existant_user_home_dir(self, reader, module,
+                                             error_class, fn_ext):
+        pytest.importorskip(module)
+
+        path = os.path.join('~', 'does_not_exist.' + fn_ext)
+        with pytest.raises(error_class):
```
Comment (WillAyd) [inline pandas/tests/io/test_common.py]: I don't think this test actually does anything within the context of this change. Is there a way to make a stronger assertion about this actually working?

Another option may be to patch a call to `_expand_user` and assert it gets called via these functions

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master and address comments

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
@pandas-dev/pandas-core if someone can merge master and update the test

**Comment by WillAyd** [comment]
Sure I'll take a look

File: pandas/tests/io/test_common.py [hunk 2]
```diff
@@ -148,6 +149,31 @@ def test_read_non_existant(self, reader, module, error_class, fn_ext):
...
+        (pd.read_json, 'os', ValueError, 'json'),
+        (pd.read_msgpack, 'os', ValueError, 'mp'),
+        (pd.read_pickle, 'os', FileNotFoundError, 'pickle'),
+    ])
+    def test_read_expands_user_home_dir(self, reader, module,
+                                             error_class, fn_ext, monkeypatch):
+        pytest.importorskip(module)
+
+        path = os.path.join('~', 'does_not_exist.' + fn_ext)
+        monkeypatch.setattr(icom, '_expand_user',
```
Comment (WillAyd) [inline pandas/tests/io/test_common.py]: Could alternately use `mock.assert_called` for this test but didn't see that library used anywhere else and I know we've been removing it from other instances, so this is an alternate way of using monkey patch to get to the same place

**Pull request merged**